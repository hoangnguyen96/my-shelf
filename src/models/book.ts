export interface BookType {
  id: string;
  title: string;
  author: string;
  category: string;
  imageUrl: string;
  description: string;
  status: boolean;
  publicationYear: number;
  rating: number;
  createdDate: string;
  edition: string;
}
