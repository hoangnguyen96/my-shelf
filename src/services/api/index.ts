import { MESSAGES } from "@app/constants";

const API_END_POINT = process.env.NEXT_PUBLIC_API_URL || "";

class ApiService {
  private _apiName: string;

  constructor(api: string) {
    this._apiName = api;
  }

  private async request<T>(
    method: string,
    url: string,
    body?: any,
    options?: RequestInit
  ) {
    const response = await fetch(`${this._apiName}${url}`, {
      method: method,
      headers: { "content-type": "application/json" },
      body: body ? JSON.stringify(body) : null,
      ...options,
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        data: undefined,
        error: new Error(MESSAGES.NETWORK_ERROR),
      };
    }

    return {
      data: data as T,
      error: undefined,
    };
  }

  async get<T>(url: string, options?: RequestInit) {
    try {
      return await this.request<T>("GET", url, null, options);
    } catch (error) {
      throw new Error(MESSAGES.GET_ERROR);
    }
  }

  async post<T>(url: string, body: any, options?: RequestInit) {
    try {
      return await this.request<T>("POST", url, body, options);
    } catch (error) {
      throw new Error(MESSAGES.POST_ERROR);
    }
  }

  async put<T>(url: string, body: any, options?: RequestInit) {
    try {
      return await this.request<T>("PUT", url, body, options);
    } catch (error) {
      throw new Error(MESSAGES.UPDATE_ERROR);
    }
  }

  async delete<T>(url: string, options?: RequestInit) {
    try {
      return await this.request<T>("DELETE", url, null, options);
    } catch (error) {
      throw new Error(MESSAGES.DELETE_ERROR);
    }
  }
}

export const api = new ApiService(API_END_POINT);
