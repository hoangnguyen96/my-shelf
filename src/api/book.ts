import { API_ROUTES } from "@app/constants";
import { BookType } from "@app/models";
import { HttpClient } from "@app/services";

export const getAllBook = async () => {
  try {
    const data = await HttpClient.get<BookType[]>(API_ROUTES.BOOKS);

    return data;
  } catch (error) {
    return [];
  }
};

export const getBookByParams = async (params: string) => {
  try {
    const data = await HttpClient.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params}`
    );

    return data;
  } catch (error) {
    return [];
  }
};

export const getBookById = async (id: number) => {
  try {
    const data = await HttpClient.get<BookType[]>(
      `${API_ROUTES.BOOKS}?id=${id}`
    );

    return data[0];
  } catch (error) {
    return [];
  }
};

export const addBook = async (payload: Partial<BookType>) => {
  return await HttpClient.post(API_ROUTES.BOOKS, payload);
};

export const updateBookById = async (
  id: string,
  payload: Partial<BookType>
) => {
  return await HttpClient.put(`${API_ROUTES.BOOKS}/${id}`, payload);
};

export const deleteBook = async (id: string) => {
  return await HttpClient.delete(`${API_ROUTES.BOOKS}/${id}`);
};
