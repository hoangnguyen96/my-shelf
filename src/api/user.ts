import { API_ROUTES } from "@app/constants";
import { User } from "@app/models";
import { HttpClient } from "@app/services";

export const getAllUser = async () => {
  const data = await HttpClient.get<User[]>(API_ROUTES.USER);

  return data || [];
};

export const getUserById = async (id: string) => {
  try {
    const data = await HttpClient.get<User[]>(
      `${API_ROUTES.USER}?userId=${id}`
    );

    return data[0];
  } catch (error) {
    return [];
  }
};

export const getUserByEmail = async (email: string) => {
  try {
    const user = (await HttpClient.get(
      `${API_ROUTES.USER}?email=${email}`
    )) as User[];

    return user;
  } catch (error) {
    return [];
  }
};

export const updateUserById = async (id: string, payload: Partial<User>) => {
  return await HttpClient.put(`${API_ROUTES.USER}/${id}`, payload);
};

export const addUser = async (payload: Partial<User>) => {
  return await HttpClient.post(API_ROUTES.USER, payload);
};
