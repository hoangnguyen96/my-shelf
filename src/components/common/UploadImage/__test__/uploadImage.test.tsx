import "@testing-library/jest-dom";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { UploadImage } from "../..";
import { User } from "@app/models";
import * as api from "@app/api";

// Mock toàn bộ module
jest.mock("@app/api", () => ({
  generateImageUpload: jest.fn(), // Mock các hàm
  updateUserById: jest.fn(),
}));

describe("UploadImage", () => {
  const mockUser: User = {
    id: "123",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    userId: "",
    username: "",
    email: "",
    password: "",
    isAdmin: false,
    phone: "",
    bio: "",
    favorites: [],
    shelfBooks: [],
  };

  beforeEach(() => {
    // Mock lại giá trị trả về trong mỗi test
    (api.generateImageUpload as jest.Mock).mockResolvedValue({
      success: true,
      data: {
        url: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
    });

    (api.updateUserById as jest.Mock).mockResolvedValue({
      success: true,
    });
  });

  it("Should render correctly snapshot", () => {
    expect(
      render(<UploadImage image={mockUser.avatar} user={mockUser} />)
    ).toMatchSnapshot();
  });

  it("should open file input when clicking upload button", () => {
    const { getByTestId, getByText } = render(
      <UploadImage image={mockUser.avatar} user={mockUser} />
    );

    const fileInput = getByTestId("fileInput");
    const uploadButton = getByText(/Upload New photo/i);
    const spyClick = jest.spyOn(fileInput, "click");

    fireEvent.click(uploadButton);

    expect(spyClick).toHaveBeenCalled();
  });
});
