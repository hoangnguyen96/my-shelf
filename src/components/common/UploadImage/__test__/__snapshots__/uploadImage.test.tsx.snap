// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage Should render correctly snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="css-0"
        enctype="multipart/form-data"
      >
        <div
          class="chakra-form-control css-1itv5e3"
          role="group"
        >
          <label
            class="chakra-form__label css-wmozf8"
            for="fileInput"
            id="myFile-label"
          >
            <img
              alt="Upload"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="110"
              src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=256&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=256&q=75 2x"
              style="color: transparent; margin: 0px auto; border-radius: 50%; width: 110px; height: 110px;"
              width="110"
            />
          </label>
          <input
            accept="image/*"
            class="chakra-input css-1hyfx7x"
            data-testid="fileInput"
            id="fileInput"
            type="file"
          />
        </div>
        <button
          class="chakra-button css-ycuso6"
          type="button"
        >
          Upload New photo
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="css-0"
      enctype="multipart/form-data"
    >
      <div
        class="chakra-form-control css-1itv5e3"
        role="group"
      >
        <label
          class="chakra-form__label css-wmozf8"
          for="fileInput"
          id="myFile-label"
        >
          <img
            alt="Upload"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="110"
            src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2F88X1WfZ%2Favatar-default.png&w=256&q=75 2x"
            style="color: transparent; margin: 0px auto; border-radius: 50%; width: 110px; height: 110px;"
            width="110"
          />
        </label>
        <input
          accept="image/*"
          class="chakra-input css-1hyfx7x"
          data-testid="fileInput"
          id="fileInput"
          type="file"
        />
      </div>
      <button
        class="chakra-button css-ycuso6"
        type="button"
      >
        Upload New photo
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
